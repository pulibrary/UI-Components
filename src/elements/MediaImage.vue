<template>
  <div class="media-image" :class="[height, {'default-thumbnail': !src}]">
    <img v-if="src" :src="src" :alt="alt" :class="[{'cover': cover}, {'contain': contain}]">
    <svg-icon v-else name="pul-icon-file" fill="rgb(225,225,225)" width="64px" height="64px"></svg-icon>
  </div>
</template>

<script>
/**
 * Cards are used to apply a container around a related grouping of information.
 */
export default {
  name: "MediaImage",
  status: "prototype",
  release: "1.0.0",
  type: "Element",
  props: {
    /**
     * The image displayed
     */
    src: {
      type: String,
      default: null,
    },
    /**
     * The alternative text describing the image. Do not include if image is decorative.
     */
    alt: {
      type: String,
      default: "",
    },
    /**
     * Manually define the height of the image for a card
     */
    height: {
      type: String,
      default: "",
    },
    /**
     * Whether the image fills the container maintaining aspect ratio and is cropped
     */
    cover: {
      type: Boolean,
      default: false,
    },
    /**
     * Whether the full image is contained within the container maintaining aspect ratio. Note that this property is not recommened for use when the `height` prop is used as it will show the background of the container.
     */
    contain: {
      type: Boolean,
      default: false,
    },
  },
}
</script>

<style lang="scss" scoped>
.media-image {
  @include reset;
  @include stack-space($space-base);
  font-family: $font-family-text;
  line-height: $line-height-base;
  color: $color-rich-black;
  overflow: hidden;
  position: relative;
  height: auto;

  img {
    background-color: $color-grayscale-lighter;
    height: auto;
    max-width: 100%;
    vertical-align: middle;
  }

  .cover {
    object-fit: cover;
  }

  .contain {
    object-fit: contain;
  }

  &.default-thumbnail {
    background: $color-grayscale;
    display: flex;
    align-items: center;

    .svg-icon,
    svg {
      display: block;
      margin: auto;
    }
  }
}

.small {
  height: 100px;
  width: 100%;

  img {
    height: 100px;
    max-width: 100%;
  }
}

.medium {
  height: 200px;
  width: 100%;

  img {
    height: 200px;
    max-width: 100%;
  }
}

.large {
  height: 300px;
  width: 100%;

  img {
    height: 300px;
    max-width: 100%;
  }
}
</style>

<docs>
  ```jsx
  <media-image src="https://picsum.photos/400/300/?random" height="medium"></media-image>
  ```
</docs>
