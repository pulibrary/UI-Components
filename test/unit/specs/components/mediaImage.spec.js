import Vuex from "vuex"
import { createLocalVue, mount, shallowMount } from "@vue/test-utils"
import MediaImage from "@/elements/MediaImage.vue"

// create an extended `Vue` constructor
const localVue = createLocalVue()

let wrapper

describe("MediaImage.vue", () => {
  beforeEach(() => {
    wrapper = mount(MediaImage, {
      localVue,
      propsData: {
        src: "",
        height: "medium",
        alt: "alt text",
        cover: true,
        contain: false,
      },
      stubs: ["svg-icon"],
    })
  })

  it("displays svg icon when src is bad or non-existent", () => {
    const mediaimg = wrapper.find(".media-image")
    expect(mediaimg.classes()).toContain("medium")
    wrapper.setProps({ height: "large" })
    expect(mediaimg.classes()).toContain("large")
    expect(mediaimg.classes()).toContain("cover")
    const svg = wrapper.find('[name="pul-icon-file"]')
    expect(wrapper.find('[name="pul-icon-file"]').exists()).toBe(true)
    expect(wrapper.find("img").exists()).toBe(false)
    expect(svg.attributes().alt).toBe("alt text")
  })

  it("displays image when src exists", () => {
    const wrapper2 = mount(MediaImage, {
      localVue,
      propsData: {
        src: "https://picsum.photos/400/300/?random",
        height: "medium",
        alt: "alt text",
        cover: false,
        contain: true,
      },
    })
    expect(wrapper2.find('[name="pul-icon-file"]').exists()).toBe(false)
    expect(wrapper2.find("img").exists()).toBe(true)
    const img = wrapper2.find("img")
    expect(img.attributes().alt).toBe("alt text")
    expect(mediaimg.classes()).toContain("contain")
  })

  it("has the expected html structure", () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
