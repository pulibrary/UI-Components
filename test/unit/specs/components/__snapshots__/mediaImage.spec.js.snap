// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaImage.vue has the expected html structure 1`] = `
<div
  class="media-image medium default-thumbnail"
>
  <svg-icon-stub
    alt="alt text"
    class="cover"
    fill="rgb(225,225,225)"
    height="64px"
    name="pul-icon-file"
    width="64px"
  />
</div>
`;
